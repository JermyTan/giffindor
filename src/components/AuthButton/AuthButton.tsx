import React, { useState } from "react";
import { GoogleLogin, GoogleLogout } from "react-google-login";
import { googleAuthClientId } from "../../config/keys";

function AuthButton() {
  const [loggedIn, setLoggedIn] = useState(false);

  const onError = (error: Error) => {
    console.log(error);
  };

  const onLoginSuccess = (response: any) => {
    console.log(response);
    setLoggedIn(true);
  };

  const onLogoutSuccess = () => {
    setLoggedIn(false);
  };

  return loggedIn ? (
    <GoogleLogout
      clientId={googleAuthClientId}
      onLogoutSuccess={onLogoutSuccess}
      onFailure={() => onError(Error("Logout error"))}
    />
  ) : (
    <GoogleLogin
      clientId={googleAuthClientId}
      onSuccess={onLoginSuccess}
      onFailure={onError}
    />
  );
}

export default AuthButton;
