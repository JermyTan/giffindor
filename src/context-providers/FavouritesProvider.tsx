import React, { createContext } from "react";
import { useFavourites } from "../utils/custom-hooks";
import { Gif } from "../components/body/gif-item/GifItem";
import { useSelector } from "react-redux";
import { getUser } from "../redux/selectors";

type FavouritesContextType = {
  isFavourite: (gifId: string) => boolean;
  favourites: Gif[];
  addToFavourites: (gifId: string) => void;
  removeFromFavourites: (gifId: string) => void;
};

export const FavouritesContext = createContext<FavouritesContextType>({
  isFavourite: (gifId: string) => false,
  favourites: [],
  addToFavourites: (gifId: string) => {},
  removeFromFavourites: (gifId: string) => {},
});

function FavouritesProvider(props: any) {
  const user = useSelector(getUser);

  const [
    favouriteGifIds,
    favourites,
    addToFavourites,
    removeFromFavourites,
  ] = useFavourites(user?.uid ?? "");

  return (
    <FavouritesContext.Provider
      value={{
        isFavourite: (gifId: string) =>
          (favouriteGifIds as Set<string>).has(gifId),
        favourites: favourites as Gif[],
        addToFavourites: addToFavourites as (gifId: string) => void,
        removeFromFavourites: removeFromFavourites as (gifId: string) => void,
      }}
    >
      {props.children}
    </FavouritesContext.Provider>
  );
}

export default FavouritesProvider;
