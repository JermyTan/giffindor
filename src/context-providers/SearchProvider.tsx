import React, { createContext, useState } from "react";

type SearchContextType = {
  searchTerm: string;
  setSearchTerm: (newSearchTerm: string) => void;
};

export const SearchContext = createContext<SearchContextType>({
  searchTerm: "",
  setSearchTerm: (newSearchTerm: string) => {},
});

function SearchProvider(props: any) {
  const [searchTerm, setSearchTerm] = useState("");

  return (
    <SearchContext.Provider
      value={{
        searchTerm: searchTerm,
        setSearchTerm: setSearchTerm,
      }}
    >
      {props.children}
    </SearchContext.Provider>
  );
}

export default SearchProvider;
