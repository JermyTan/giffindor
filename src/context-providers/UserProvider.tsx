import React, { createContext, useState, useContext } from "react";
import { FavouritesContext } from "./FavouritesProvider";
import { provider, auth } from "../utils/firebase";

type User = {
  uid: string;
  displayName: string;
};

type UserContextType = {
  user: User | null;
  login: () => void;
  logout: () => void;
  setUser: (user: User | null) => void;
};

export const UserContext = createContext<UserContextType>({
  user: null,
  login: () => {},
  logout: () => {},
  setUser: (user: User | null) => {},
});

function UserProvider(props: any) {
  const { setShowFavourites } = useContext(FavouritesContext);
  const [user, setUser] = useState<User | null>(null);

  const login = () => {
    auth
      .signInWithPopup(provider)
      .then((response) => {
        const uid = response.user?.uid ?? "";
        const displayName = response.user?.displayName ?? "";
        const user =
          uid && displayName
            ? {
                uid: uid,
                displayName: displayName,
              }
            : null;

        setUser(user);
      })
      .catch((error) => {
        console.log(error);
      });
  };

  const logout = () => {
    auth
      .signOut()
      .then(() => {
        setUser(null);
        setShowFavourites(false);
      })
      .catch((error) => {
        console.log(error);
      });
  };

  return (
    <UserContext.Provider
      value={{
        user: user,
        login: login,
        logout: logout,
        setUser: setUser,
      }}
    >
      {props.children}
    </UserContext.Provider>
  );
}

export default UserProvider;
