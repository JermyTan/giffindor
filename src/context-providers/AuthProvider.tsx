import React, { createContext, useContext } from "react";
import { FavouritesContext } from "./FavouritesProvider";
import { UserContext } from "./UserProvider";
import { provider, auth } from "../utils/firebase";

type AuthContextType = {
  login: () => void;
  logout: () => void;
};

export const AuthContext = createContext<AuthContextType>({
  login: () => {},
  logout: () => {},
});

function AuthProvider(props: any) {
  const { setUser } = useContext(UserContext);
  const { setShowFavourites } = useContext(FavouritesContext);

  const login = () => {
    auth
      .signInWithPopup(provider)
      .then((response) => {
        const uid = response.user?.uid ?? "";
        const displayName = response.user?.displayName ?? "";
        const user =
          uid && displayName
            ? {
                uid: uid,
                displayName: displayName,
              }
            : null;

        setUser(user);
      })
      .catch((error) => {
        console.log(error);
      });
  };

  const logout = () => {
    auth
      .signOut()
      .then(() => {
        setUser(null);
        setShowFavourites(false);
      })
      .catch((error) => {
        console.log(error);
      });
  };

  return (
    <AuthContext.Provider value={{ login: login, logout: logout }}>
      {props.children}
    </AuthContext.Provider>
  );
}

export default AuthProvider;
