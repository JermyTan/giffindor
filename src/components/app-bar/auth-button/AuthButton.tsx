import React, { useContext } from "react";
import { Button } from "semantic-ui-react";
import { UserContext } from "../../../context-providers/UserProvider";
import { AuthContext } from "../../../context-providers/AuthProvider";

type Props = {
  style?: any;
  className?: string;
};

function AuthButton(props: Props) {
  const { user } = useContext(UserContext);
  const { login, logout } = useContext(AuthContext);

  return (
    <Button
      data-testid="auth-button"
      content={user ? "Logout" : "Login"}
      onClick={user ? logout : login}
      style={props.style}
      className={props.className}
    />
  );
}

export default AuthButton;
