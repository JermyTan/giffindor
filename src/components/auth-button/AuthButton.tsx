import React, { useContext } from "react";
import { Button } from "semantic-ui-react";
import { provider, auth } from "../../firebase";
import { UserContext } from "../../context-providers/UserProvider";

type Props = {
  style?: any;
  className?: string;
};

function AuthButton(props: Props) {
  const userContext = useContext(UserContext);

  const onError = (error: Error) => {
    console.log(error);
  };

  const onLogin = () => {
    auth
      .signInWithPopup(provider)
      .then((response) => {
        const uid = response.user?.uid ?? "";
        const displayName = response.user?.displayName ?? "";
        const user =
          uid && displayName
            ? {
                uid: uid,
                displayName: displayName,
              }
            : null;

        userContext.setUser(user);
      })
      .catch(onError);
  };

  const onLogout = () => {
    auth
      .signOut()
      .then(() => {
        userContext.setUser(null);
      })
      .catch(onError);
  };

  return userContext.user ? (
    <Button onClick={onLogout} style={props.style} className={props.className}>
      Logout
    </Button>
  ) : (
    <Button onClick={onLogin} style={props.style} className={props.className}>
      Login
    </Button>
  );
}

export default AuthButton;
