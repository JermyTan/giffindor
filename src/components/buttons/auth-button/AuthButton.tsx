import React, { useContext } from "react";
import { Button } from "semantic-ui-react";
import { UserContext } from "../../../context-providers/UserProvider";

type Props = {
  style?: any;
  className?: string;
};

function AuthButton(props: Props) {
  const { user, login, logout } = useContext(UserContext);

  return (
    <Button
      content={user ? "Logout" : "Login"}
      onClick={user ? logout : login}
      style={props.style}
      className={props.className}
    />
  );
}

export default AuthButton;
