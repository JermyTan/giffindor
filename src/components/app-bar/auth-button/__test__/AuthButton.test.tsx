import React from "react";
import ReactDom from "react-dom";
import AuthButton from "../AuthButton";
import { render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { UserContext } from "../../../../context-providers/UserProvider";
import { AuthContext } from "../../../../context-providers/AuthProvider";
import { create } from "react-test-renderer";

afterEach(cleanup);

function generateAuthButton(loggedIn: boolean = false) {
  // create mock
  const login = jest.fn();
  const logout = jest.fn();
  const mockUser = { uid: "123", displayName: "Jeremy" };

  const authButton = (
    <UserContext.Provider
      value={{ user: loggedIn ? mockUser : null, setUser: jest.fn() }}
    >
      <AuthContext.Provider value={{ login: login, logout: logout }}>
        <AuthButton className="header-field" />
      </AuthContext.Provider>
    </UserContext.Provider>
  );

  return {
    authButton: authButton,
    login: login,
    logout: logout,
  };
}

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDom.render(<AuthButton />, div);

  const authButtonWithLoggedOutState = generateAuthButton().authButton;
  ReactDom.render(authButtonWithLoggedOutState, div);

  const authButtonWithLoggedInState = generateAuthButton(true).authButton;
  ReactDom.render(authButtonWithLoggedInState, div);
});

it("renders login state correctly", () => {
  const { authButton } = generateAuthButton();
  const { getByTestId } = render(authButton);

  expect(getByTestId("auth-button")).toHaveTextContent("Login");
});

it("renders logout state correctly", () => {
  const { authButton } = generateAuthButton(true);
  const { getByTestId } = render(authButton);

  expect(getByTestId("auth-button")).toHaveTextContent("Logout");
});

it("calls login correctly", () => {
  const { authButton, login, logout } = generateAuthButton();
  const tree = create(authButton);

  tree.root.findByType("button").props.onClick();

  expect(login).toHaveBeenCalledTimes(1);
  expect(logout).toHaveBeenCalledTimes(0);
});

it("calls logout correctly", () => {
  const { authButton, login, logout } = generateAuthButton(true);
  const tree = create(authButton);

  tree.root.findByType("button").props.onClick();

  expect(login).toHaveBeenCalledTimes(0);
  expect(logout).toHaveBeenCalledTimes(1);
});

it("matches login snapshot", () => {
  const { authButton } = generateAuthButton();
  const tree = create(authButton).toJSON();

  expect(tree).toMatchSnapshot();
});

it("matches logout snapshot", () => {
  const { authButton } = generateAuthButton(true);
  const tree = create(authButton).toJSON();

  expect(tree).toMatchSnapshot();
});
