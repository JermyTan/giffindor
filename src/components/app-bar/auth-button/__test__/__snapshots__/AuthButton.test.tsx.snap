// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches login snapshot 1`] = `
<button
  className="ui button header-field"
  data-testid="auth-button"
  onClick={[Function]}
>
  Login
</button>
`;

exports[`matches logout snapshot 1`] = `
<button
  className="ui button header-field"
  data-testid="auth-button"
  onClick={[Function]}
>
  Logout
</button>
`;
